redfish-core/lib/assembly.hpp

```diff
@@ -0,0 +1,495 @@
+            asyncResp->res.jsonValue["Assemblies"];
+        asyncResp->res.jsonValue["Assemblies@odata.count"] =
+            assemblyArray.size();
+
+        assemblyIndex++;
+    }
+}
+
+/**
+ * @brief Api to check if the assemblies fetched from association Json is
+ * also implemented in the system. In case the interface for that assembly
+ * is not found update the list and fetch properties for only implemented
+ * assemblies.
+ * @param[in] asyncResp - Shared pointer for asynchronous calls.
+ * @param[in] chassisPath - Chassis the assemblies are associated with.
+ * @param[in] assemblies - list of all the assemblies associated with the
+ * chassis.
+ * @return None.
+ */
+inline void
+    checkAssemblyInterface(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                           const std::string& chassisPath,
+                           const dbus::utility::MapperEndPoints& assemblies)
+{
+    constexpr std::array<std::string_view, 5> fruInterface = {
+        "xyz.openbmc_project.Inventory.Item.Vrm",
+        "xyz.openbmc_project.Inventory.Item.Tpm",
+        "xyz.openbmc_project.Inventory.Item.Panel",
+        "xyz.openbmc_project.Inventory.Item.Battery",
+        "xyz.openbmc_project.Inventory.Item.DiskBackplane"};
+
+    dbus::utility::getSubTree(
+        "/xyz/openbmc_project/inventory", 0, fruInterface,
+        [asyncResp, chassisPath,
+         assemblies](const boost::system::error_code& ec,
+                     const dbus::utility::MapperGetSubTreeResponse& subtree) {
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR("D-Bus response error on GetSubTree {}",
+                             ec.value());
+            messages::internalError(asyncResp->res);
+            return;
+        }
+
+        if (subtree.empty())
+        {
+            BMCWEB_LOG_DEBUG("No object paths found");
+            return;
+        }
+        std::vector<std::string> updatedAssemblyList;
+        for (const auto& [objectPath, serviceName] : subtree)
+        {
+            // This list will store common paths between assemblies fetched
+            // from association json and assemblies which are actually
+            // implemeted. This is to handle the case in which there is
+            // entry in association json but implementation of interface for
+            // that particular assembly is missing.
+            auto it = std::find(assemblies.begin(), assemblies.end(),
+                                objectPath);
+            if (it != assemblies.end())
+            {
+                updatedAssemblyList.emplace(updatedAssemblyList.end(), *it);
+            }
+        }
+
+        if (!updatedAssemblyList.empty())
+        {
+            // sorting is required to facilitate patch as the array does not
+            // store and data which can be mapped back to Dbus path of
+            // assembly.
+            std::sort(updatedAssemblyList.begin(), updatedAssemblyList.end());
```

Comment:
Upstream suggests to use `ranges` - https://github.com/openbmc/bmcweb/commit/3544d2a719c8bb7b07f9a39f61a3770ec84b909f

```
 std::ranges::sort(updatedAssemblyList);
```